hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala

```diff
@@ -28,19 +28,188 @@ case class TableStage(
     CollectDistributedArray(contexts, broadcastVals, "context", "global", bodyTransform(body))
 
   def broadcastRef: IR = Ref("global", broadcastVals.typ)
+
   def contextRef: IR = Ref("context", contextType)
+
   def globals: IR = GetField(broadcastRef, globalsField)
 }
 
 object LowerTableIR {
-  def lower(ir: IR, typesToLower: DArrayLowering.Type): IR =
-    new LowerTableIR(typesToLower).lower(ir)
-}
+  def apply(ir: IR, typesToLower: DArrayLowering.Type, ctx: ExecuteContext): IR = {
```

Comment:
I worked very hard elsewhere to avoid this elsewhere.  I do not think optimization compiler transformation passes should be doing IO.  That should only happen when creating Read/er nodes on construction.

Reply:
this isn't doing IO...

I'm adding the Execute context here for the timer and the Backend.

Reply:
(this is a precursor to the shuffle lowering infra)

Reply:
Ah, sorry, I got mixed up. The table native reader stuff below does IO, and I just rewrote it all, but that hasn't landed and this is how it was before.  Nevermind.

Reply:
But, for the record: IR nodes shouldn't do IO except on construction.

Reply:
yes, agreed.